Third-party-libraries.md

```diff
@@ -0,0 +1,145 @@
+* `third_party/python_libs` stores our production Python dependencies.
+* `third_party/static` stores some of our production Node.js dependencies.
+* `node_modules` stores more production Node.js dependencies. On developer machines, it also stores development dependencies.
+* `third_party/` also stores some other production dependencies like protobuf files.
+
+When you run `python -m scripts.start`, a chain of scripts installs and/or upgrades dependencies as necessary:
+
+```text
+          +----------+
+          | start.py |
+          +----------+
+               |
+               | calls install_third_party_libs.main() when start.py is executed or imported
+               v
+  +-----------------------------+
+  | install_third_party_libs.py |
+  +-----------------------------+
+               |
+               | main() calls install_third_party.main()
+               v
+    +------------------------+
+    | install_third_party.py |
+    +------------------------+
+               |
+               | main() calls install_backend_python_libs.main()
+               v
++--------------------------------+
+| install_backend_python_libs.py |
++--------------------------------+
+```
+
+We'll look at each of these scripts below.
+
+### `start.py`
+
+When you start a development server, you execute `scripts/start.py`. This file does not itself install any dependencies, but it contains a call to `install_third_party_libs.main()` that will be executed whenever `start.py` is executed, or even imported.
+
+### `install_third_party_libs.py`
+
+#### Python packages
+
+Whenever `install_third_party_libs.py` is executed or imported, it installs some Python packages that are needed early by the installation process. Some are installed to `oppia_tools/`, and other are installed to `third_party/python_libs/`.
+
+When `install_third_party_libs.main()` runs, it installs additional Python packages:
+
+* Packages that need to be available on the system `PATH` are installed "globally" using pip (in practice these are installed into your virtual environment).
+* Other packages are installed to `oppia_tools/`.
+
+All of these Python packages are listed directly in the `install_third_party_libs.py` file.
+
+For all the Python packages installed by this script, the version of the package is pinned (i.e. we always install the same version, even if a newer version is available). However, the dependencies of that package may not have pinned versions. For example, we depend on `pip-tools`, which in turn depends on `pip`. The `setup.py` in `pip-tools` does not specify a particular version of `pip` that it requires, so when we install `pip-tools`, we will install the latest version of `pip`. This can lead to test failures when a breaking change in `pip` is released.
```

Comment:
Hm ... we should fix this?

More generally, should we try to pin all the things?

Reply:
Yes, we should pin everything. @vojtechjelinek is working on this as part of moving our Python libs to a virtual environment